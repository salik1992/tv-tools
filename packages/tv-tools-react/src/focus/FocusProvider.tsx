import type { ComponentProps, FocusEvent, KeyboardEvent } from 'react';
import { useCallback, useMemo } from 'react';
import { isDirectional } from '@salik1992/tv-tools/control';
import { FocusManager } from '@salik1992/tv-tools/focus';
import { noop } from '@salik1992/tv-tools/utils/noop';
import { FocusContext } from './context';

/**
 * The main root component for the focus management.
 * It creates a div element that is used for listening to key events in the manager and
 * it also creates the focus manager instance for the application which is available through
 * the FocusContext or useFocusManager hook.
 * Use at the root of your application.
 * @prop alwaysPreventNavigationalEvents - Use this to prevent TV use their own navigation.
 * @prop ...props - Any other props that you want to pass to the wrapping div element.
 */
export const FocusProvider = ({
	children,
	alwaysPreventNavigationalEvents,
	onFocus: onOuterFocus,
	onKeyDownCapture: onOuterKeyDownCapture,
	onKeyDown: onOuterKeyDown,
	onKeyUpCapture: onOuterKeyUpCapture,
	onKeyUp: onOuterKeyUp,
	...props
}: ComponentProps<'div'> & {
	/** Use this to prevent TV use their own navigation. */
	alwaysPreventNavigationalEvents?: boolean;
}) => {
	const focusManager = useMemo(() => new FocusManager(), []);

	const onFocus = useCallback(
		(event: FocusEvent<HTMLDivElement>) => {
			focusManager.handleFocusEvent(event.nativeEvent);
			onOuterFocus?.(event);
		},
		[focusManager, onOuterFocus],
	);

	const keyDownCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'capture', event);
			onOuterKeyDownCapture?.(event);
		},
		[focusManager, onOuterKeyDownCapture],
	);

	const keyDownBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keydown', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyDown?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyDown],
	);

	const keyUpCapture = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'capture', event);
			onOuterKeyUpCapture?.(event);
		},
		[focusManager, onOuterKeyUpCapture],
	);

	const keyUpBubble = useCallback(
		(event: KeyboardEvent<HTMLDivElement>) => {
			focusManager.handleKeyEvent('keyup', 'bubble', event);
			if (
				alwaysPreventNavigationalEvents &&
				isDirectional(event) &&
				document.activeElement?.tagName !== 'INPUT'
			) {
				event.preventDefault();
				event.stopPropagation();
			} else {
				onOuterKeyUp?.(event);
			}
		},
		[focusManager, alwaysPreventNavigationalEvents, onOuterKeyUp],
	);

	const focusContextValue = useMemo(
		() => ({
			focusManager,
			addChild: noop,
		}),
		[focusManager],
	);

	return (
		<div
			{...props}
			onFocus={onFocus}
			onKeyDownCapture={keyDownCapture}
			onKeyDown={keyDownBubble}
			onKeyUpCapture={keyUpCapture}
			onKeyUp={keyUpBubble}
		>
			<FocusContext.Provider value={focusContextValue}>
				{children}
			</FocusContext.Provider>
		</div>
	);
};
