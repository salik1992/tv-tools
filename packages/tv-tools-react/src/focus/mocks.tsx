import type { RefObject } from 'react';
import { useContext } from 'react';
import type { FocusManager } from '@salik1992/tv-tools/focus';
import { FocusContext } from './context';

export const ExposeFocusManager = ({
	focusManager,
}: {
	focusManager: RefObject<FocusManager | null>;
}) => {
	const context = useContext(FocusContext);
	if (context.focusManager) {
		focusManager.current = context.focusManager;
	}
	return null;
};

export function assertFocusManager(
	focusManager: RefObject<FocusManager | null>,
): asserts focusManager is RefObject<FocusManager> {
	if (focusManager.current === null) {
		throw new Error('FocusManager is null');
	}
}
