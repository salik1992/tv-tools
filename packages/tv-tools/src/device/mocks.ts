import type { Feature } from './base';

export const Features: Feature[] = [
	'volume',
	'screensaver',
	'4k',
	'8k',
	'multiplayer',
	'widevine-ctr',
	'widevine-cbc',
	'playready-ctr',
	'playready-cbc',
	'fairplay',
] as const;
