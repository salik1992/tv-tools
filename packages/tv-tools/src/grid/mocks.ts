import { toKeys } from '../utils/toKeys';
import type { RenderData } from './types';

export const assertRenderData = (
	renderData: RenderData<string>,
	focusSpy: jest.SpyInstance,
	assertion: Partial<{
		focus: string;
		baseOffset: number;
		nextArrow: boolean;
		previousArrow: boolean;
		ids: [string, string[]][];
		dataIndices: number[][];
		offsets: number[];
	}>,
) => {
	for (const key of toKeys(assertion)) {
		switch (key) {
			case 'focus':
				if (assertion.focus) {
					expect(focusSpy).toHaveBeenCalledWith(assertion.focus, {
						preventScroll: true,
					});
				}
				break;
			case 'baseOffset':
				if (assertion.baseOffset !== renderData.baseOffset) {
					throw new Error(
						`Expected baseOffset to be ${assertion.baseOffset} but got ${renderData.baseOffset}`,
					);
				}
				break;
			case 'nextArrow':
				if (assertion.nextArrow !== renderData.nextArrow) {
					throw new Error(
						`Expected nextArrow to be ${assertion.nextArrow} but got ${renderData.nextArrow}`,
					);
				}
				break;
			case 'previousArrow':
				if (assertion.previousArrow !== renderData.previousArrow) {
					throw new Error(
						`Expected previousArrow to be ${assertion.previousArrow} but got ${renderData.previousArrow}`,
					);
				}
				break;
			case 'ids':
				if (assertion.ids) {
					for (let i = 0; i < assertion.ids.length; i++) {
						if (assertion.ids[i][0] !== renderData.groups[i].id) {
							throw new Error(
								`Expected group id ${assertion.ids[i][0]} at index ${i} but got ${renderData.groups[i].id}`,
							);
						}
						for (let j = 0; j < assertion.ids[i][1].length; j++) {
							if (
								assertion.ids[i][1][j] !==
								renderData.groups[i].elements[j].id
							) {
								throw new Error(
									`Expected element id ${assertion.ids[i][1][j]} at index ${j} but got ${renderData.groups[i].elements[j].id}`,
								);
							}
						}
					}
				}
				break;
			case 'dataIndices':
				if (assertion.dataIndices) {
					for (let i = 0; i < assertion.dataIndices.length; i++) {
						for (
							let j = 0;
							j < assertion.dataIndices[i].length;
							j++
						) {
							if (
								assertion.dataIndices[i][j] !==
								renderData.groups[i].elements[j].dataIndex
							) {
								throw new Error(
									`Expected dataIndex ${assertion.dataIndices[i][j]} at index ${j} but got ${renderData.groups[i].elements[j].dataIndex}`,
								);
							}
						}
					}
				}
				break;
			case 'offsets':
				if (assertion.offsets) {
					for (let i = 0; i < assertion.offsets.length; i++) {
						if (
							assertion.offsets[i] !== renderData.groups[i].offset
						) {
							throw new Error(
								`Expected offset ${assertion.offsets[i]} at index ${i} but got ${renderData.groups[i].offset}`,
							);
						}
					}
					break;
				}
		}
		focusSpy.mockClear();
	}
};
