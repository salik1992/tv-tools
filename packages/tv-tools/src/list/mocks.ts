import { toKeys } from '../utils/toKeys';
import type { RenderData } from './types';

export const assertRenderData = (
	renderData: RenderData<string>,
	focusSpy: jest.SpyInstance,
	assertion: Partial<{
		focus: string;
		baseOffset: number;
		nextArrow: boolean;
		previousArrow: boolean;
		ids: string[];
		dataIndices: number[];
		offsets: number[];
	}>,
) => {
	for (const key of toKeys(assertion)) {
		switch (key) {
			case 'focus':
				if (assertion.focus) {
					expect(focusSpy).toHaveBeenCalledWith(assertion.focus, {
						preventScroll: true,
					});
				}
				break;
			case 'baseOffset':
				if (assertion.baseOffset !== renderData.baseOffset) {
					console.log('throwing');
					throw new Error(
						`Expected baseOffset to be ${assertion.baseOffset} but got ${renderData.baseOffset}`,
					);
				}
				break;
			case 'nextArrow':
				if (assertion.nextArrow !== renderData.nextArrow) {
					throw new Error(
						`Expected nextArrow to be ${assertion.nextArrow} but got ${renderData.nextArrow}`,
					);
				}
				break;
			case 'previousArrow':
				if (assertion.previousArrow !== renderData.previousArrow) {
					throw new Error(
						`Expected previousArrow to be ${assertion.previousArrow} but got ${renderData.previousArrow}`,
					);
				}
				break;
			case 'ids':
				if (assertion.ids) {
					for (let i = 0; i < assertion.ids.length; i++) {
						if (assertion.ids[i] !== renderData.elements[i].id) {
							throw new Error(
								`Expected id ${assertion.ids[i]} at index ${i} but got ${renderData.elements[i].id}`,
							);
						}
					}
				}
				break;
			case 'dataIndices':
				if (assertion.dataIndices) {
					for (let i = 0; i < assertion.dataIndices.length; i++) {
						if (
							assertion.dataIndices[i] !==
							renderData.elements[i].dataIndex
						) {
							console.log(
								assertion.dataIndices,
								renderData.elements.map(
									({ dataIndex }) => dataIndex,
								),
							);
							throw new Error(
								`Expected dataIndex ${assertion.dataIndices[i]} at index ${i} but got ${renderData.elements[i].dataIndex}`,
							);
						}
					}
				}
				break;
			case 'offsets':
				if (assertion.offsets) {
					for (let i = 0; i < assertion.offsets.length; i++) {
						if (
							assertion.offsets[i] !==
							renderData.elements[i].offset
						) {
							throw new Error(
								`Expected offset ${assertion.offsets[i]} at index ${i} but got ${renderData.elements[i].offset}`,
							);
						}
					}
				}
				break;
		}
	}
	focusSpy.mockClear();
};
