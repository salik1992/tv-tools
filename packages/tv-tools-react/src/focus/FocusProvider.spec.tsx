import { createRef } from 'react';
import { render } from '@testing-library/react';
import type { FocusManager } from '@salik1992/tv-tools/focus';
import { FocusProvider } from './FocusProvider';
import { ExposeFocusManager, assertFocusManager } from './mocks';

const NAVIGATIONAL_CODE = 'ArrowDown';
const OTHER_CODE = 'Enter';

describe('FocusProvider', () => {
	const focusManager = createRef<FocusManager>();

	it('should render the children', () => {
		const { container } = render(
			<FocusProvider>
				<div>Test</div>
				<ExposeFocusManager focusManager={focusManager} />
			</FocusProvider>,
		);
		expect(container.innerHTML).toMatchSnapshot();
	});

	it('should pass keyboard events to the focus manager', () => {
		const { container } = render(
			<FocusProvider id="focusRoot">
				<div>Test</div>
				<ExposeFocusManager focusManager={focusManager} />
			</FocusProvider>,
		);
		const keydDownBubble = new KeyboardEvent('keydown', {
			code: 'ArrowDown',
			bubbles: true,
		});
		const stopSpyDB = jest.spyOn(keydDownBubble, 'stopPropagation');
		const keyUpBubble = new KeyboardEvent('keyup', {
			code: 'ArrowDown',
			bubbles: true,
		});
		const stopSpyUB = jest.spyOn(keyUpBubble, 'stopPropagation');
		assertFocusManager(focusManager);
		const spy = jest.spyOn(focusManager.current, 'handleKeyEvent');
		const root = container.querySelector('#focusRoot');
		root?.dispatchEvent(keydDownBubble);
		root?.dispatchEvent(keyUpBubble);
		expect(spy).toHaveBeenNthCalledWith(
			1,
			'keydown',
			'capture',
			expect.objectContaining({ code: NAVIGATIONAL_CODE }),
		);
		expect(spy).toHaveBeenNthCalledWith(
			2,
			'keydown',
			'bubble',
			expect.objectContaining({ code: NAVIGATIONAL_CODE }),
		);
		expect(spy).toHaveBeenNthCalledWith(
			3,
			'keyup',
			'capture',
			expect.objectContaining({ code: NAVIGATIONAL_CODE }),
		);
		expect(spy).toHaveBeenNthCalledWith(
			4,
			'keyup',
			'bubble',
			expect.objectContaining({ code: NAVIGATIONAL_CODE }),
		);
		expect(stopSpyDB).not.toHaveBeenCalled();
		expect(stopSpyUB).not.toHaveBeenCalled();
	});

	it('should prevent default and stop propagation for navigational events with alwaysPreventNavigationalEvents', () => {
		const { container } = render(
			<FocusProvider id="focusRoot" alwaysPreventNavigationalEvents>
				<div>Test</div>
				<ExposeFocusManager focusManager={focusManager} />
			</FocusProvider>,
		);
		const keydDownBubble = new KeyboardEvent('keydown', {
			code: 'ArrowDown',
			bubbles: true,
		});
		const preventSpyDB = jest.spyOn(keydDownBubble, 'preventDefault');
		const stopSpyDB = jest.spyOn(keydDownBubble, 'stopPropagation');
		const keyUpBubble = new KeyboardEvent('keyup', {
			code: 'ArrowDown',
			bubbles: true,
		});
		const preventSpyUB = jest.spyOn(keyUpBubble, 'preventDefault');
		const stopSpyUB = jest.spyOn(keyUpBubble, 'stopPropagation');
		assertFocusManager(focusManager);
		const spy = jest.spyOn(focusManager.current, 'handleKeyEvent');
		const root = container.querySelector('#focusRoot');
		root?.dispatchEvent(keydDownBubble);
		root?.dispatchEvent(keyUpBubble);
		expect(spy).toHaveBeenNthCalledWith(
			1,
			'keydown',
			'capture',
			expect.objectContaining({ code: NAVIGATIONAL_CODE }),
		);
		expect(spy).toHaveBeenNthCalledWith(
			2,
			'keydown',
			'bubble',
			expect.objectContaining({ code: NAVIGATIONAL_CODE }),
		);
		expect(spy).toHaveBeenNthCalledWith(
			3,
			'keyup',
			'capture',
			expect.objectContaining({ code: NAVIGATIONAL_CODE }),
		);
		expect(spy).toHaveBeenNthCalledWith(
			4,
			'keyup',
			'bubble',
			expect.objectContaining({ code: NAVIGATIONAL_CODE }),
		);
		expect(preventSpyDB).toHaveBeenCalled();
		expect(stopSpyDB).toHaveBeenCalled();
		expect(preventSpyUB).toHaveBeenCalled();
		expect(stopSpyUB).toHaveBeenCalled();
	});

	it('should not prevent default and stop propagation for non-navigational events with alwaysPreventNavigationalEvents', () => {
		const { container } = render(
			<FocusProvider id="focusRoot" alwaysPreventNavigationalEvents>
				<div>Test</div>
				<ExposeFocusManager focusManager={focusManager} />
			</FocusProvider>,
		);
		const keydDownBubble = new KeyboardEvent('keydown', {
			code: OTHER_CODE,
			bubbles: true,
		});
		const preventSpyDB = jest.spyOn(keydDownBubble, 'preventDefault');
		const stopSpyDB = jest.spyOn(keydDownBubble, 'stopPropagation');
		const keyUpBubble = new KeyboardEvent('keyup', {
			code: OTHER_CODE,
			bubbles: true,
		});
		const preventSpyUB = jest.spyOn(keyUpBubble, 'preventDefault');
		const stopSpyUB = jest.spyOn(keyUpBubble, 'stopPropagation');
		assertFocusManager(focusManager);
		const spy = jest.spyOn(focusManager.current, 'handleKeyEvent');
		const root = container.querySelector('#focusRoot');
		root?.dispatchEvent(keydDownBubble);
		root?.dispatchEvent(keyUpBubble);
		expect(spy).toHaveBeenNthCalledWith(
			1,
			'keydown',
			'capture',
			expect.objectContaining({ code: OTHER_CODE }),
		);
		expect(spy).toHaveBeenNthCalledWith(
			2,
			'keydown',
			'bubble',
			expect.objectContaining({ code: OTHER_CODE }),
		);
		expect(spy).toHaveBeenNthCalledWith(
			3,
			'keyup',
			'capture',
			expect.objectContaining({ code: OTHER_CODE }),
		);
		expect(spy).toHaveBeenNthCalledWith(
			4,
			'keyup',
			'bubble',
			expect.objectContaining({ code: OTHER_CODE }),
		);
		expect(preventSpyDB).not.toHaveBeenCalled();
		expect(stopSpyDB).not.toHaveBeenCalled();
		expect(preventSpyUB).not.toHaveBeenCalled();
		expect(stopSpyUB).not.toHaveBeenCalled();
	});
});
