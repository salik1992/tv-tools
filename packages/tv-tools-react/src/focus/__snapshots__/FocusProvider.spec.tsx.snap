// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FocusProvider should render the children 1`] = `"<div><div>Test</div></div>"`;
